### PR TITLE
Refactor the implementation of Goos

### LINKED ISSUE #364: Grasshopper_Toolkit: GH_IBHoMGeometry should specify  GH_GeometricGoo<IGeometry>
Instead of  `GH_GeometricGoo<object>`.

We have now separated the rhino entity and the bhom entity in the Geometry.

`Value` and `m_Value` always refer to a `BH.oM.Geometry.IGeometry` object
`m_RhinoGeometry` works in the background and always refers to a `Rhino.Geometry` object

### LINKED ISSUE #365: Grasshopper_Toolkit: GH_BHoMObject should specify GH_BHoMGoo<BHoMObject>
instead of GH_BHoMGoo<object>.

Following the amendement that brought to #364.
At the moment you can pass anything in `GH_BHoMObject` which is conflicting with its base definition. The reflection here is not symmetric, and not that transparent.
e.g. `IGeometry` is not a `BHoMObject` and should fail, whereas it currently succeeds.

Passing a `IGeometry` would work in `GH_IObject` though.

### PR DESCRIPTION
**Pull request by alelom**
### Issues addressed by this PR

Closes #364
Closes #365

This pr is clarifying the behaviour and centralises the logic of the BHoM `Goo`s.
Especially:
- `IGeometryGoo` is now specialised for `IGeometry`ies
- `IObjectGoo` is now specialised for `IObject`s
- `BHoMObjectGoo` is now specialised for `BHoMObject`s 

### Changelog

- `GH_IBHoMGeometry` now specifies `GH_GeometricGoo<IGeometry>` instead of `object`
- `GH_BHoMObject` now specifies `GH_BHoMGoo<BHoMObject>` instead of `object`
- All the `Goo`s now have more explicit names. E.g. `GH_Enum` is `EnumGoo`
- All `Goo`s now can be baked, if any geometry is embedded in them
- Conversion from framework types to `Goo`s has been centralised and is now more robust
- Prevent anunhandled exception in `Type.CastFrom` when the source is a string. This is a pre-fix, see https://github.com/BHoM/BHoM_Engine/issues/1193 for a permanent fix.
- Added automatic casting from CustomObject to Dictionary and vice versa

### CODE REVIEW

**Comment by FraserGreenroyd** [comment]
Hi @alelom please can you fix the title of this PR. It should be a single sentence describing the changes for the change log to be generated and not have `GH_Toolkit` in the title :smile:

**Comment by epignatelli** [comment]
Hey @alelom I'll be in today from lunch. I'll put something into the calendar to move this forward!

File: Grasshopper_Engine/Objects/Types/GH_BHoMObject.cs [hunk 1]
```diff
@@ -36,58 +36,38 @@
 
 namespace BH.Engine.Grasshopper.Objects
 {
-    public class GH_BHoMObject : GH_BHoMGoo<object>, IGH_PreviewData, GH_ISerializable
+    public class BHoMObjectGoo : GH_BHoMGoo<BHoMObject>, IGH_PreviewData, GH_ISerializable
```
Comment (epignatelli) [inline Grasshopper_Engine/Objects/Types/GH_BHoMObject.cs]: Shouldn't this be `IBHoMObject` according to what you told me last time?

File: Grasshopper_Engine/Compute/RenderRhinoMesh.cs [hunk 2]
```diff
@@ -52,11 +52,11 @@ public static void IRenderRhinoMeshes(this object geometry, GH_PreviewMeshArgs a
 
         /***************************************************/
 
-        public static void RenderRhinoMeshes(this List<object> geometry, Rhino.Display.DisplayPipeline pipeline, DisplayMaterial material)
+        public static void IRenderRhinoMeshes(this List<object> geometry, Rhino.Display.DisplayPipeline pipeline, DisplayMaterial material)
```
[the remark this answers is not in the document]
Comment (epignatelli) [reply Grasshopper_Engine/Compute/RenderRhinoMesh.cs]: Superseeded  by the last commits - resolving

**Comment by epignatelli** [comment]
One thing that I would do only after we stop modifying code for the purpose of this pr is to rename the files that host the BHoM `Goo`s with their respective class names. That way we keep the diff clear on what's changed, otherwise we get all greens and reds.

File: Grasshopper_Engine/Objects/Types/GH_BHoMObject.cs (hunk 1 above)
Comment (alelom) [reply Grasshopper_Engine/Objects/Types/GH_BHoMObject.cs]: You're right, simply forgot to change it! Thanks

File: Grasshopper_Engine/Objects/Types/GH_IBHoMGeometry.cs [hunk 3]
```diff
@@ -200,35 +197,35 @@ public override bool CastFrom(object source)
...
+                Value = transform.Value.ToBHoM();
             }
             else if (source is Rhino.Geometry.Ellipse)
             {
-                m_Value = ((Rhino.Geometry.Ellipse)source).ToBHoM();
+                Value = ((Rhino.Geometry.Ellipse)source).ToBHoM();
             }
             else
-                m_Value = GH_Convert.ToGeometryBase(source).IToBHoM();
+                Value = GH_Convert.ToGeometryBase(source).IToBHoM();
```
Comment (epignatelli) [inline Grasshopper_Engine/Objects/Types/GH_IBHoMGeometry.cs]: Now an `IGeometryGoo` does not accept all the types that accepted before, e.g. `System` types,  custom libraries types.

I think that this is ok for now, but we should record an error and turn the parameter red when this happens.
At the state of this pr, the parameter stays blank and returns null, which is not very descriptive.

This is true also for the other two type we have modified: `BHoMObjectGoo` and `IObjectGoo`.

**Comment by epignatelli** [comment]
From the Sprint Closure call @peterjamesnugent and @IsakNaslundBh offered to use a couple scripts to test this.

**Comment by FraserGreenroyd** [comment]
Apologies @epignatelli this week ran off from me at the end here, I'll take a look on Monday early doors if @IsakNaslundBh or @peterjamesnugent are unable to before then @smile:

**Comment by peterjamesnugent** [comment]
Tested two different scripts:
- You can longer cast from `BH.oM.Geometry.Point` to `Point`, this gives the following error:
>  1. Data conversion failed from IGeometry to Point

- This was useful to make use of native functions such as creating a set to remove duplicates and retrieve their indexes

I tried opening the Truss_Model script in \samples\Structural_Adapters\Lusas_Toolkit\Workflows but get the following error: 
> 1. Impossible to convert BH.oM.Adapters.Lusas.Split1D into BH.Engine.Grasshopper.Objects.EnumGoo. Check the input description for more details on the type of object that need to be provided

**Comment by alelom** [comment]
I think @peterjamesnugent 's points are valid -- need to have a deeper look at this with @epignatelli when he's back at the end of the week

**Comment by epignatelli** [comment]
Thanks all for the reviews!

@peterjamesnugent I fixed the issues - well spotted!

There is one major issue. The `BHoMAdapter` cannot be held by any of our parameters. Specifically it is neither a `BHoMObject` nor an `IObject`.

If everything is alright with the rest, I see two solutions:
A. We merge, and we use the "Data" parameter that grasshopper provides by default.
We then can: a) Create a BHoMAdapter parameter in grasshopper; b) Make BHoMAdapter inherit from `IObject` or `BHoMObject`; c) Keep using the data parameter.
I prefer b)

B. We wait for the merge and check the Adapter situation from the framework point of view first.

It both A. and B, we still have alternatives and do not block the usage of grasshopper.

**Comment by peterjamesnugent** [comment]
### Truss Model - Lusas_Toolkit Samples
This error still appears in the [Truss_Model ](https://github.com/BHoM/samples/tree/master/Structural_Adapters/Lusas_Toolkit/Workflows) on start up:

image

I don't think it is the C# Script component that is causing the issue, it's the component prior to it, which has the following error:

> 1. Recursive data stream found, this component depends on itself

### Large Bridge Script

From a fairly large script I encounter the following errors:

> 1. Cannot cast BH.oM.Geometry.Line to IBHoMObject

This one is fairly self explanatory.

> 1. Data conversion failed from Goo to IBHoMObject

This is the error from sending a `BHoMAdapter` to one of the BHoM Parameters. 

I think we should address the `BHoMAdapter` issue as all of our samples use the BHoM Parameter and it has been something we have advocated as best practice through training and samples.

**Comment by alelom** [comment]
> > 1. Data conversion failed from Goo to IBHoMObject
> 
> This is the error from sending a `BHoMAdapter` to one of the BHoM Parameters.
> 
> I think we should address the `BHoMAdapter` issue as all of our samples use the BHoM Parameter and it has been something we have advocated as best practice through training and samples.

This one is addressed [here ](https://github.com/BHoM/BHoM_Adapter/issues/131) and we plan to solve it ASAP.

**Comment by peterjamesnugent** [comment]
@alelom I think we should create a branch and a draft PR for that issue and make this PR dependent on it being merged. We don't want this PR to be merged without it and risk having an installer(s) created that is not compatible with scripts (such as samples) before and after this PR is merged.

It would probably affect a small amount of people but it wouldn't be good practice.

**Comment by alelom** [comment]
> > 1. Recursive data stream found, this component depends on itself

@epignatelli this one is indeed due to an issue in the casting of `BH.Engine.Grasshopper.Objects.EnumGoo`:
image

**Comment by peterjamesnugent** [comment]
Similar issue for faces:

> 1. Cannot cast BH.oM.Geometry.Face to IBHoMObject

**Comment by epignatelli** [comment]
@peterjamesnugent do I need to input something in the script to generate the error? 
It doesn't pop up to me.

**Comment by peterjamesnugent** [comment]
> @peterjamesnugent do I need to input something in the script to generate the error?
> It doesn't pop up to me.

For Issue 1 the recursive error comes from the component creating a 1D Mesh:
image

**Comment by alelom** [comment]
@peterjamesnugent works for me now. Coud you re-test it too?

**Comment by peterjamesnugent** [comment]
I no longer get the error, I am unable to push objects but I don't think that is related to this PR.

**Comment by epignatelli** [comment]
Thanks @peterjamesnugent ! Do the objects get pushed if you use the master branch?

**Comment by peterjamesnugent** [comment]
Same problem on master branch.

**Comment by epignatelli** [comment]
Ah, ok, that's a relief.

So, to summarise, the only problem that still stays is the one for the `BHoMAdapter `.

Following a conversation with @al-fisher, I am going to push a new `BHoMAdapterParam`, regardless of  the outcome of https://github.com/BHoM/BHoM_Adapter/issues/131.

This way we should have everything on the plate for merging.

**Comment by al-fisher** [comment]
Thanks @epignatelli

**Comment by alelom** [comment]
Sounds good, thanks @epignatelli!

**Comment by epignatelli** [comment]
Before I push the new adapter, here is a quick utility that replaces the old parameters with the new ones. 
It will not make its way into the 2.4 milestone because of time, but will be there in the 2.5 MVP.
I'll open an issue with the code to make this more visible for the future.

[ReplaceWithNewParams.zip]
(https://github.com/BHoM/Grasshopper_Toolkit/files/3632196/ReplaceWithNewParams.zip)

**Comment by epignatelli** [comment]
Alright guys!
I pushed the `BHoMAdapter_Param`, so we can now hold it somewhere. 

Remember that it will be annoying to open the scripts and find things red - so do check out [this utility](https://github.com/BHoM/Grasshopper_Toolkit/files/3632196/ReplaceWithNewParams.zip) to replace the parameters. You need to:
1. Input the type of the new parameter
1. Select the ones you want to replace
1. Hit the button
1. That's it.

@peterjamesnugent @FraserGreenroyd if the pr is ok for you now, would you mind giving me a green flag, so I can merge?

**Comment by al-fisher** [comment]
To be merged end of Friday/over weekend. @peterjamesnugent subject to you confirming and approving! Ta

**Comment by adecler** [comment]
This PR has too much impact to be merged so close to the release date. I have moved it to the next milestone as it is not critical.

**Comment by epignatelli** [comment]
Thanks @adecler for taking care of blocking it - agreed. I don't feel comfortable merging this either.

I'll close, and keep the branch for the 2.5 RC milestone (it will require a brand new branch anyway).

**Comment by peterjamesnugent** [comment]
Added a section on the Grasshopper_Toolkit Wiki that relates to this refactoring, so we should update it when this PR gets reraised: 

https://github.com/BHoM/Grasshopper_Toolkit/wiki/Internalising-BHoMObjects

**Pull request closed without merging**